src/core/smp.cc

```diff
@@ -218,17 +219,27 @@ internal::memory_prefaulter::memory_prefaulter(const resource::resources& res, m
             }
             a.set(cpuset);
         }
-        _worker_threads.emplace_back(a, [this, &ranges, page_size, huge_page_size_opt] {
+        _worker_threads.emplace_back(a, [this, &alien, &ranges, page_size, huge_page_size_opt] {
+            ++_active_threads;
             work(ranges, page_size, huge_page_size_opt);
+            if (!--_active_threads) {
+                run_on(alien, 0, [this] () noexcept { join_threads(); });
```

Comment:
Is there no risk that this `join_threads()` call will happen after the `memory_prefaulter` is already destroyed?

Reply:
Yes, but's it's not nice to have implicit lifetime assumptions like that.
Not that this case is worth caring about, though.

Reply:
In such cases it is unavoidable without strong compiler support. The setup is too hairy.